### PR TITLE
Add a dry-run to verify placement fields

### PR DESCRIPTION
**Pull request by omeryahud**
**What this PR does / why we need it**:
This PR adds a deployment creation dry-run to the SSP webhook to reject any invalid placement fields.

**Which issue(s) this PR fixes**:
Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1938241

**Special notes for your reviewer**:
- Still missing tests

**Release note**:

```release-note
None
```

### CODE REVIEW

File: api/v1beta1/ssp_webhook.go [hunk 1]
```diff
@@ -65,6 +69,11 @@ func (r *SSP) ValidateCreate() error {
 		return fmt.Errorf("creation failed, the configured namespace for common templates does not exist: %v", namespaceName)
 	}
 
+	err = validatePlacement(r)
```
Comment (marceloamaral) [inline api/v1beta1/ssp_webhook.go:72]: For simplicity and coding size, I like the error testing in only one line:

if err = validatePlacement(r); err != nil {
	return err
}

File: api/v1beta1/ssp_webhook.go [hunk 2]
```diff
@@ -91,3 +105,71 @@ func (r *SSP) ValidateDelete() error {
...
+	err = validateOperandPlacement(ssp.Spec.TemplateValidator.Placement)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func validateOperandPlacement(placement *api.NodePlacement) error {
+	if placement == nil {
```
Comment (marceloamaral) [inline api/v1beta1/ssp_webhook.go:124]: Shouldn't this return an error saying that there is not placement rules to be applied?

Comment (omeryahud) [reply api/v1beta1/ssp_webhook.go:124]: No. No placement rules is acceptable, it basically tells the k8s scheduler that we don't care where our workloads are placed, we just want them running

File: api/v1beta1/ssp_webhook.go (hunk 1 above)
Comment (omeryahud) [reply api/v1beta1/ssp_webhook.go]: Done

File: api/v1beta1/ssp_webhook.go [hunk 3]
```diff
@@ -91,3 +104,67 @@ func (r *SSP) ValidateDelete() error {
 func setClientForWebhook(c client.Client) {
 	clt = c
 }
+
+func validatePlacement(ssp *SSP) error {
+	if err := validateOperandPlacement(ssp.Spec.NodeLabeller.Placement); err != nil {
```
Comment (ksimon1) [inline api/v1beta1/ssp_webhook.go:109]: This is no longer needed

File: tests/webhook_test.go [hunk 4]
```diff
@@ -39,55 +87,131 @@ var _ = Describe("Validation webhook", func() {
...
 				}
 				Expect(err.Error()).To(ContainSubstring(fmt.Sprintf(
 					"creation failed, the configured namespace for common templates does not exist: %v",
-					foundSsp.Spec.CommonTemplates.Namespace,
+					newSsp.Spec.CommonTemplates.Namespace,
 				)))
 			})
+
+			Context("Placement API validation", func() {
+				Context("node-labeller", func() {
```
Comment (ksimon1) [inline tests/webhook_test.go:128]: This is no longer needed

File: api/v1beta1/ssp_webhook.go (hunk 3 above)
Comment (omeryahud) [reply api/v1beta1/ssp_webhook.go]: Removed

File: tests/webhook_test.go (hunk 4 above)
Comment (omeryahud) [reply tests/webhook_test.go]: Removed

File: api/v1beta1/ssp_webhook.go [hunk 5]
```diff
@@ -91,3 +104,63 @@ func (r *SSP) ValidateDelete() error {
...
 	clt = c
 }
+
+func validatePlacement(ssp *SSP) error {
+	if err := validateOperandPlacement(ssp.Spec.TemplateValidator.Placement); err != nil {
+		return err
+	}
+
+	return nil
+}
```
Comment (akrejcir) [inline api/v1beta1/ssp_webhook.go:114]: This function can be inlined, or simplified to:
```
return validateOperandPlacement(ssp.Spec.TemplateValidator.Placement)
```

File: tests/webhook_test.go [hunk 6]
```diff
@@ -39,55 +87,100 @@ var _ = Describe("Validation webhook", func() {
...
 				}
 				Expect(err.Error()).To(ContainSubstring(fmt.Sprintf(
 					"creation failed, the configured namespace for common templates does not exist: %v",
-					foundSsp.Spec.CommonTemplates.Namespace,
+					newSsp.Spec.CommonTemplates.Namespace,
 				)))
 			})
+
+			Context("Placement API validation", func() {
+				Context("template-validator", func() {
```
Comment (akrejcir) [inline tests/webhook_test.go:128]: This second `Context` is not needed. Each `It` test case already mentions `template-validator`.

File: tests/webhook_test.go [hunk 7]
```diff
@@ -39,55 +87,100 @@ var _ = Describe("Validation webhook", func() {
...
-			foundSsp := getSsp()
 			originalNs := foundSsp.Spec.CommonTemplates.Namespace
 			foundSsp.Spec.CommonTemplates.Namespace = originalNs + "-updated"
 			err := apiClient.Update(ctx, foundSsp)
 			Expect(err).To(HaveOccurred())
 			Expect(err.Error()).To(ContainSubstring("commonTemplates.namespace cannot be changed."))
 		})
+
+		Context("Placement API validation", func() {
+			Context("template-validator", func() {
```
Comment (akrejcir) [inline tests/webhook_test.go:171]: The same comment as above.

**Review by akrejcir — COMMENTED** [review]
Looks good. I only have a few optional code style comments. You don't have to change them if you want to save time.

/lgtm

Is this still on hold? The test case is there.

File: api/v1beta1/ssp_webhook.go (hunk 5 above)
Comment (omeryahud) [reply api/v1beta1/ssp_webhook.go:114]: Yes, I agree, I will change that

**Pull request merged**